INDI/INDI/indiserver.c

```diff
@@ -838,46 +824,57 @@ static void shutdownDvr(DvrInfo *dp, int restart)
         /* close local named FIFOs */
         close(dp->wfd);
         close(dp->rfd);
-#       if 0
-        close(dp->efd);
-#       endif/*0*/
     }
 
 #ifdef OSX_EMBEDED_MODE
     fprintf(stderr, "STOPPED \"%s\"\n", dp->name);
     fflush(stderr);
 #endif
 
-    /* free memory */
-    free(dp->sprops);
-    free(dp->dev);
-    delLilXML(dp->lp);
+    /* free memory; ensure no double-free if stop interrupts restart */
+    if (dp->sprops) { free(dp->sprops); dp->sprops = NULL; }
+    if (dp->dev) { free(dp->dev); dp->dev = NULL; }
+    if (dp->lp) { delLilXML(dp->lp); dp->lp = NULL; }
 
     /* ok now to recycle */
     dp->active = 0;
-    dp->ndev   = 0;
 
     /* decrement and possibly free any unsent messages for this client */
-    while ((mp = (Msg *)popFQ(dp->msgq)) != NULL)
-        if (--mp->count == 0)
-            freeMsg(mp);
-    delFQ(dp->msgq);
+    if (dp->msgq)
+    {
+        while ((mp = (Msg *)popFQ(dp->msgq)) != NULL)
+        {
+            if (--mp->count == 0) { freeMsg(mp); }
+        }
+        delFQ(dp->msgq);
+        dp->msgq = NULL;
+    }
 
     if (restart)
     {
-        if (dp->restarts >= maxrestarts)
+        if (dp->restarts >= maxrestarts && maxrestarts > 0)
```

Comment:
If maxrestarts is zero or less, then the number of restart attempts is unlimited.